File: ExelBidDemoObjC/ExelBidDemoObjC/DesignSystem.m
#import "DesignSystem.h"

const CGFloat EBSpacingXS  = 4;
const CGFloat EBSpacingS   = 8;
const CGFloat EBSpacingM   = 12;
const CGFloat EBSpacingL   = 16;
const CGFloat EBSpacingXL  = 20;
const CGFloat EBSpacingXXL = 28;

const CGFloat EBCardCornerRadius   = 14;
const CGFloat EBButtonCornerRadius = 12;

UIColor *EBBrandAccentColor(void) {
    return UIColor.systemBlueColor;
}

static UIFont *EBFontWithWeight(UIFont *font, UIFontWeight weight) {
    UIFontDescriptor *descriptor = [font.fontDescriptor fontDescriptorByAddingAttributes:@{
        UIFontDescriptorTraitsAttribute: @{UIFontWeightTrait: @(weight)}
    }];
    return [UIFont fontWithDescriptor:descriptor size:0];
}

#pragma mark - EBCardView

@interface EBCardView ()
@property (nonatomic, strong) UIStackView *stack;
@end

@implementation EBCardView

- (instancetype)init {
    return [self initWithSpacing:EBSpacingM
                          insets:UIEdgeInsetsMake(16, 16, 16, 16)];
}

- (instancetype)initWithSpacing:(CGFloat)spacing insets:(UIEdgeInsets)insets {
    self = [super initWithFrame:CGRectZero];
    if (!self) { return nil; }

    self.backgroundColor = UIColor.secondarySystemGroupedBackgroundColor;
    self.layer.cornerRadius = EBCardCornerRadius;
    self.layer.cornerCurve = kCACornerCurveContinuous;

    _stack = [[UIStackView alloc] init];
    _stack.axis = UILayoutConstraintAxisVertical;
    _stack.spacing = spacing;
    _stack.alignment = UIStackViewAlignmentFill;
    _stack.translatesAutoresizingMaskIntoConstraints = NO;
    [self addSubview:_stack];

    [NSLayoutConstraint activateConstraints:@[
        [_stack.topAnchor      constraintEqualToAnchor:self.topAnchor      constant:insets.top],
        [_stack.bottomAnchor   constraintEqualToAnchor:self.bottomAnchor   constant:-insets.bottom],
        [_stack.leadingAnchor  constraintEqualToAnchor:self.leadingAnchor  constant:insets.left],
        [_stack.trailingAnchor constraintEqualToAnchor:self.trailingAnchor constant:-insets.right]
    ]];
    return self;
}

- (void)addArranged:(UIView *)view {
    [self.stack addArrangedSubview:view];
}

- (void)addArrangedSubviews:(NSArray<UIView *> *)views {
    for (UIView *v in views) { [self.stack addArrangedSubview:v]; }
}

- (void)setCustomSpacing:(CGFloat)spacing afterView:(UIView *)view {
    [self.stack setCustomSpacing:spacing afterView:view];
}

@end

#pragma mark - EBSectionLabel

@implementation EBSectionLabel
+ (UILabel *)make:(NSString *)text {
    UILabel *l = [[UILabel alloc] init];
    l.text = text.uppercaseString;
    l.font = EBFontWithWeight([UIFont preferredFontForTextStyle:UIFontTextStyleCaption1], UIFontWeightSemibold);
    l.textColor = UIColor.secondaryLabelColor;
    l.adjustsFontForContentSizeCategory = YES;
    return l;
}
@end

#pragma mark - EBTypography

@implementation EBTypography

+ (UILabel *)title:(NSString *)text {
    UILabel *l = [[UILabel alloc] init];
    l.text = text;
    l.font = EBFontWithWeight([UIFont preferredFontForTextStyle:UIFontTextStyleTitle2], UIFontWeightBold);
    l.numberOfLines = 0;
    l.adjustsFontForContentSizeCategory = YES;
    return l;
}

+ (UILabel *)body:(NSString *)text {
    UILabel *l = [[UILabel alloc] init];
    l.text = text;
    l.font = [UIFont preferredFontForTextStyle:UIFontTextStyleBody];
    l.textColor = UIColor.labelColor;
    l.numberOfLines = 0;
    l.adjustsFontForContentSizeCategory = YES;
    return l;
}

+ (UILabel *)footnote:(NSString *)text {
    UILabel *l = [[UILabel alloc] init];
    l.text = text;
    l.font = [UIFont preferredFontForTextStyle:UIFontTextStyleFootnote];
    l.textColor = UIColor.secondaryLabelColor;
    l.numberOfLines = 0;
    l.adjustsFontForContentSizeCategory = YES;
    return l;
}

@end

#pragma mark - EBInfoRow

@interface EBInfoRow ()
@property (nonatomic, strong) UILabel *valueLabel;
@end

@implementation EBInfoRow

- (instancetype)initWithKey:(NSString *)key value:(NSString *)value {
    self = [super initWithFrame:CGRectZero];
    if (!self) { return nil; }

    UILabel *keyLabel = [[UILabel alloc] init];
    keyLabel.text = key;
    keyLabel.font = [UIFont preferredFontForTextStyle:UIFontTextStyleSubheadline];
    keyLabel.textColor = UIColor.secondaryLabelColor;
    keyLabel.adjustsFontForContentSizeCategory = YES;
    [keyLabel setContentHuggingPriority:UILayoutPriorityRequired forAxis:UILayoutConstraintAxisHorizontal];

    _valueLabel = [[UILabel alloc] init];
    _valueLabel.text = value;
    _valueLabel.font = [UIFont preferredFontForTextStyle:UIFontTextStyleSubheadline];
    _valueLabel.textColor = UIColor.labelColor;
    _valueLabel.textAlignment = NSTextAlignmentRight;
    _valueLabel.numberOfLines = 0;
    _valueLabel.adjustsFontForContentSizeCategory = YES;

    UIStackView *stack = [[UIStackView alloc] initWithArrangedSubviews:@[keyLabel, _valueLabel]];
    stack.axis = UILayoutConstraintAxisHorizontal;
    stack.spacing = EBSpacingM;
    stack.alignment = UIStackViewAlignmentFirstBaseline;
    stack.translatesAutoresizingMaskIntoConstraints = NO;
    [self addSubview:stack];
    [NSLayoutConstraint activateConstraints:@[
        [stack.topAnchor      constraintEqualToAnchor:self.topAnchor],
        [stack.bottomAnchor   constraintEqualToAnchor:self.bottomAnchor],
        [stack.leadingAnchor  constraintEqualToAnchor:self.leadingAnchor],
        [stack.trailingAnchor constraintEqualToAnchor:self.trailingAnchor]
    ]];
    return self;
}

- (void)setValue:(NSString *)text { self.valueLabel.text = text; }

@end

#pragma mark - EBPrimaryButton

@implementation EBPrimaryButton

+ (UIButton *)filled:(NSString *)title target:(id)target action:(SEL)action {
    UIButton *b = [UIButton buttonWithType:UIButtonTypeCustom];
    [b setTitle:title forState:UIControlStateNormal];
    [b setTitleColor:UIColor.whiteColor forState:UIControlStateNormal];
    [b setTitleColor:[UIColor.whiteColor colorWithAlphaComponent:0.5] forState:UIControlStateDisabled];
    b.titleLabel.font = EBFontWithWeight([UIFont preferredFontForTextStyle:UIFontTextStyleBody], UIFontWeightSemibold);
    b.titleLabel.adjustsFontForContentSizeCategory = YES;
    b.backgroundColor = EBBrandAccentColor();
    b.layer.cornerRadius = EBButtonCornerRadius;
    b.layer.cornerCurve = kCACornerCurveContinuous;
    b.contentEdgeInsets = UIEdgeInsetsMake(10, 16, 10, 16);
    if (target && action) {
        [b addTarget:target action:action forControlEvents:UIControlEventTouchUpInside];
    }
    return b;
}

+ (UIButton *)tinted:(NSString *)title target:(id)target action:(SEL)action {
    UIButton *b = [UIButton buttonWithType:UIButtonTypeCustom];
    UIColor *accent = EBBrandAccentColor();
    [b setTitle:title forState:UIControlStateNormal];
    [b setTitleColor:accent forState:UIControlStateNormal];
    [b setTitleColor:[accent colorWithAlphaComponent:0.4] forState:UIControlStateDisabled];
    b.titleLabel.font = EBFontWithWeight([UIFont preferredFontForTextStyle:UIFontTextStyleBody], UIFontWeightSemibold);
    b.titleLabel.adjustsFontForContentSizeCategory = YES;
    b.backgroundColor = [accent colorWithAlphaComponent:0.12];
    b.layer.cornerRadius = EBButtonCornerRadius;
    b.layer.cornerCurve = kCACornerCurveContinuous;
    b.contentEdgeInsets = UIEdgeInsetsMake(10, 16, 10, 16);
    if (target && action) {
        [b addTarget:target action:action forControlEvents:UIControlEventTouchUpInside];
    }
    return b;
}

@end

#pragma mark - EBSurfaceCardView

@interface EBSurfaceCardView ()
@property (nonatomic, copy) void (^action)(void);
@end

@implementation EBSurfaceCardView

- (instancetype)initWithTitle:(NSString *)title
                     subtitle:(NSString *)subtitle
                       symbol:(NSString *)symbolName
                       action:(void (^)(void))action {
    self = [super initWithFrame:CGRectZero];
    if (!self) { return nil; }
    _action = [action copy];

    self.backgroundColor = UIColor.secondarySystemGroupedBackgroundColor;
    self.layer.cornerRadius = EBCardCornerRadius;
    self.layer.cornerCurve = kCACornerCurveContinuous;

    UIImageView *icon = [[UIImageView alloc] initWithImage:[UIImage systemImageNamed:symbolName]];
    icon.tintColor = EBBrandAccentColor();
    icon.contentMode = UIViewContentModeScaleAspectFit;
    [icon setContentHuggingPriority:UILayoutPriorityRequired forAxis:UILayoutConstraintAxisHorizontal];

    UILabel *titleLabel = [[UILabel alloc] init];
    titleLabel.text = title;
    titleLabel.font = [UIFont preferredFontForTextStyle:UIFontTextStyleHeadline];
    titleLabel.adjustsFontForContentSizeCategory = YES;

    UILabel *subtitleLabel = [[UILabel alloc] init];
    subtitleLabel.text = subtitle;
    subtitleLabel.font = [UIFont preferredFontForTextStyle:UIFontTextStyleFootnote];
    subtitleLabel.textColor = UIColor.secondaryLabelColor;
    subtitleLabel.numberOfLines = 0;
    subtitleLabel.adjustsFontForContentSizeCategory = YES;

    UIStackView *textStack = [[UIStackView alloc] initWithArrangedSubviews:@[titleLabel, subtitleLabel]];
    textStack.axis = UILayoutConstraintAxisVertical;
    textStack.spacing = 2;

    UIImageView *chevron = [[UIImageView alloc] initWithImage:[UIImage systemImageNamed:@"chevron.right"]];
    chevron.tintColor = UIColor.tertiaryLabelColor;
    [chevron setContentHuggingPriority:UILayoutPriorityRequired forAxis:UILayoutConstraintAxisHorizontal];

    UIStackView *row = [[UIStackView alloc] initWithArrangedSubviews:@[icon, textStack, chevron]];
    row.axis = UILayoutConstraintAxisHorizontal;
    row.spacing = EBSpacingM;
    row.alignment = UIStackViewAlignmentCenter;
    row.userInteractionEnabled = NO;
    row.translatesAutoresizingMaskIntoConstraints = NO;
    [self addSubview:row];

    [NSLayoutConstraint activateConstraints:@[
        [row.topAnchor      constraintEqualToAnchor:self.topAnchor      constant:14],
        [row.bottomAnchor   constraintEqualToAnchor:self.bottomAnchor   constant:-14],
        [row.leadingAnchor  constraintEqualToAnchor:self.leadingAnchor  constant:16],
        [row.trailingAnchor constraintEqualToAnchor:self.trailingAnchor constant:-16],
        [icon.widthAnchor   constraintEqualToConstant:28],
        [icon.heightAnchor  constraintEqualToConstant:28]
    ]];

    UIButton *tap = [UIButton buttonWithType:UIButtonTypeSystem];
    tap.translatesAutoresizingMaskIntoConstraints = NO;
    [tap addTarget:self action:@selector(handleTap) forControlEvents:UIControlEventTouchUpInside];
    [self addSubview:tap];
    [NSLayoutConstraint activateConstraints:@[
        [tap.topAnchor      constraintEqualToAnchor:self.topAnchor],
        [tap.bottomAnchor   constraintEqualToAnchor:self.bottomAnchor],
        [tap.leadingAnchor  constraintEqualToAnchor:self.leadingAnchor],
        [tap.trailingAnchor constraintEqualToAnchor:self.trailingAnchor]
    ]];
    return self;
}

- (void)handleTap { if (self.action) { self.action(); } }

@end

#pragma mark - EBLogView

@interface EBLogView ()
@property (nonatomic, strong) UITextView *textView;
@end

@implementation EBLogView

- (instancetype)initWithFrame:(CGRect)frame {
    self = [super initWithFrame:frame];
    if (!self) { return nil; }
    self.backgroundColor = UIColor.systemGroupedBackgroundColor;
    self.layer.cornerRadius = 8;
    self.layer.cornerCurve = kCACornerCurveContinuous;
    self.clipsToBounds = YES;

    _textView = [[UITextView alloc] init];
    _textView.font = [UIFont monospacedSystemFontOfSize:11 weight:UIFontWeightRegular];
    _textView.textColor = UIColor.labelColor;
    _textView.backgroundColor = UIColor.clearColor;
    _textView.editable = NO;
    _textView.alwaysBounceVertical = YES;
    _textView.textContainerInset = UIEdgeInsetsMake(8, 8, 8, 8);
    _textView.translatesAutoresizingMaskIntoConstraints = NO;
    [self addSubview:_textView];
    [NSLayoutConstraint activateConstraints:@[
        [_textView.topAnchor      constraintEqualToAnchor:self.topAnchor],
        [_textView.bottomAnchor   constraintEqualToAnchor:self.bottomAnchor],
        [_textView.leadingAnchor  constraintEqualToAnchor:self.leadingAnchor],
        [_textView.trailingAnchor constraintEqualToAnchor:self.trailingAnchor],
        [_textView.heightAnchor   constraintGreaterThanOrEqualToConstant:160]
    ]];
    return self;
}

- (void)append:(NSString *)line {
    NSString *stamp = [NSDateFormatter localizedStringFromDate:[NSDate date]
                                                     dateStyle:NSDateFormatterNoStyle
                                                     timeStyle:NSDateFormatterMediumStyle];
    NSString *text = self.textView.text ?: @"";
    self.textView.text = [text stringByAppendingFormat:@"[%@] %@\n", stamp, line];
    NSRange end = NSMakeRange(self.textView.text.length, 0);
    [self.textView scrollRangeToVisible:end];
}

- (void)clear { self.textView.text = @""; }

@end

File: ExelBidDemo/ExelBidDemo/DesignSystem.swift
import UIKit

/// Shared design tokens + small UIKit primitives. Targets the HIG defaults
/// (system colors, dynamic type, secondarySystemGroupedBackground cards)
/// while staying compatible with the demo's iOS 13 deployment target.
enum Spacing {
    static let xs: CGFloat  = 4
    static let s: CGFloat   = 8
    static let m: CGFloat   = 12
    static let l: CGFloat   = 16
    static let xl: CGFloat  = 20
    static let xxl: CGFloat = 28
}

enum CornerRadius {
    static let card: CGFloat   = 14
    static let button: CGFloat = 12
}

enum BrandColor {
    static let accent = UIColor.systemBlue
}

/// Container that visually groups related controls. Use `addArranged(_:)`
/// to stack subviews vertically inside the card with the standard padding.
final class CardView: UIView {

    private let stack = UIStackView()

    init(spacing: CGFloat = Spacing.m, insets: UIEdgeInsets = .init(top: 16, left: 16, bottom: 16, right: 16)) {
        super.init(frame: .zero)
        backgroundColor = .secondarySystemGroupedBackground
        layer.cornerRadius = CornerRadius.card
        layer.cornerCurve = .continuous

        stack.axis = .vertical
        stack.spacing = spacing
        stack.alignment = .fill
        stack.translatesAutoresizingMaskIntoConstraints = false
        addSubview(stack)
        NSLayoutConstraint.activate([
            stack.topAnchor.constraint(equalTo: topAnchor, constant: insets.top),
            stack.bottomAnchor.constraint(equalTo: bottomAnchor, constant: -insets.bottom),
            stack.leadingAnchor.constraint(equalTo: leadingAnchor, constant: insets.left),
            stack.trailingAnchor.constraint(equalTo: trailingAnchor, constant: -insets.right)
        ])
    }

    @available(*, unavailable)
    required init?(coder: NSCoder) { fatalError() }

    func addArranged(_ view: UIView) {
        stack.addArrangedSubview(view)
    }

    func addArranged(_ views: [UIView]) {
        views.forEach { stack.addArrangedSubview($0) }
    }

    func setCustomSpacing(_ s: CGFloat, after view: UIView) {
        stack.setCustomSpacing(s, after: view)
    }
}

/// Section/eyebrow label — small, uppercase, tertiary color.
enum SectionLabel {
    static func make(_ text: String) -> UILabel {
        let l = UILabel()
        l.text = text.uppercased()
        l.font = UIFont.preferredFont(forTextStyle: .caption1).withWeight(.semibold)
        l.textColor = .secondaryLabel
        l.adjustsFontForContentSizeCategory = true
        return l
    }
}

enum Typography {
    static func title(_ text: String) -> UILabel {
        let l = UILabel()
        l.text = text
        l.font = UIFont.preferredFont(forTextStyle: .title2).withWeight(.bold)
        l.numberOfLines = 0
        l.adjustsFontForContentSizeCategory = true
        return l
    }

    static func body(_ text: String) -> UILabel {
        let l = UILabel()
        l.text = text
        l.font = UIFont.preferredFont(forTextStyle: .body)
        l.textColor = .label
        l.numberOfLines = 0
        l.adjustsFontForContentSizeCategory = true
        return l
    }

    static func footnote(_ text: String) -> UILabel {
        let l = UILabel()
        l.text = text
        l.font = UIFont.preferredFont(forTextStyle: .footnote)
        l.textColor = .secondaryLabel
        l.numberOfLines = 0
        l.adjustsFontForContentSizeCategory = true
        return l
    }
}

/// `key — value` two-column row used in info / status cards.
final class InfoRow: UIView {

    let valueLabel = UILabel()
    private let keyLabel = UILabel()

    init(key: String, value: String = "") {
        super.init(frame: .zero)

        keyLabel.text = key
        keyLabel.font = .preferredFont(forTextStyle: .subheadline)
        keyLabel.textColor = .secondaryLabel
        keyLabel.setContentHuggingPriority(.required, for: .horizontal)
        keyLabel.adjustsFontForContentSizeCategory = true

        valueLabel.text = value
        valueLabel.font = .preferredFont(forTextStyle: .subheadline)
        valueLabel.textColor = .label
        valueLabel.textAlignment = .right
        valueLabel.numberOfLines = 0
        valueLabel.adjustsFontForContentSizeCategory = true

        let stack = UIStackView(arrangedSubviews: [keyLabel, valueLabel])
        stack.axis = .horizontal
        stack.spacing = Spacing.m
        stack.alignment = .firstBaseline
        stack.translatesAutoresizingMaskIntoConstraints = false
        addSubview(stack)
        NSLayoutConstraint.activate([
            stack.topAnchor.constraint(equalTo: topAnchor),
            stack.bottomAnchor.constraint(equalTo: bottomAnchor),
            stack.leadingAnchor.constraint(equalTo: leadingAnchor),
            stack.trailingAnchor.constraint(equalTo: trailingAnchor)
        ])
    }

    @available(*, unavailable)
    required init?(coder: NSCoder) { fatalError() }

    func set(_ text: String) { valueLabel.text = text }
}

/// Primary / secondary button factory. Manual styling so it works on
/// iOS 13+ (UIButton.Configuration would need iOS 15).
enum PrimaryButton {
    static func filled(_ title: String, target: Any?, action: Selector) -> UIButton {
        let b = UIButton(type: .custom)
        b.setTitle(title, for: .normal)
        b.setTitleColor(.white, for: .normal)
        b.setTitleColor(UIColor.white.withAlphaComponent(0.5), for: .disabled)
        b.titleLabel?.font = UIFont.preferredFont(forTextStyle: .body).withWeight(.semibold)
        b.titleLabel?.adjustsFontForContentSizeCategory = true
        b.backgroundColor = BrandColor.accent
        b.layer.cornerRadius = CornerRadius.button
        b.layer.cornerCurve = .continuous
        b.contentEdgeInsets = UIEdgeInsets(top: 10, left: 16, bottom: 10, right: 16)
        b.addTarget(target, action: action, for: .touchUpInside)
        return b
    }

    static func tinted(_ title: String, target: Any?, action: Selector) -> UIButton {
        let b = UIButton(type: .custom)
        b.setTitle(title, for: .normal)
        b.setTitleColor(BrandColor.accent, for: .normal)
        b.setTitleColor(BrandColor.accent.withAlphaComponent(0.4), for: .disabled)
        b.titleLabel?.font = UIFont.preferredFont(forTextStyle: .body).withWeight(.semibold)
        b.titleLabel?.adjustsFontForContentSizeCategory = true
        b.backgroundColor = BrandColor.accent.withAlphaComponent(0.12)
        b.layer.cornerRadius = CornerRadius.button
        b.layer.cornerCurve = .continuous
        b.contentEdgeInsets = UIEdgeInsets(top: 10, left: 16, bottom: 10, right: 16)
        b.addTarget(target, action: action, for: .touchUpInside)
        return b
    }
}

extension UIFont {
    func withWeight(_ weight: UIFont.Weight) -> UIFont {
        let descriptor = fontDescriptor.addingAttributes([
            .traits: [UIFontDescriptor.TraitKey.weight: weight]
        ])
        return UIFont(descriptor: descriptor, size: 0)
    }
}

/// Tappable card used by the Home / Ads / Mediation / MPartners list
/// screens. Icon + title + subtitle + optional trailing badge + chevron.
final class SurfaceCardView: UIView {

    private let action: () -> Void

    init(
        title: String,
        subtitle: String,
        symbol: String,
        accessoryText: String? = nil,
        action: @escaping () -> Void
    ) {
        self.action = action
        super.init(frame: .zero)

        backgroundColor = .secondarySystemGroupedBackground
        layer.cornerRadius = CornerRadius.card
        layer.cornerCurve = .continuous

        let icon = UIImageView(image: UIImage(systemName: symbol))
        icon.tintColor = BrandColor.accent
        icon.contentMode = .scaleAspectFit
        icon.setContentHuggingPriority(.required, for: .horizontal)

        let titleLabel = UILabel()
        titleLabel.text = title
        titleLabel.font = UIFont.preferredFont(forTextStyle: .headline)
        titleLabel.adjustsFontForContentSizeCategory = true

        let subtitleLabel = UILabel()
        subtitleLabel.text = subtitle
        subtitleLabel.font = .preferredFont(forTextStyle: .footnote)
        subtitleLabel.textColor = .secondaryLabel
        subtitleLabel.numberOfLines = 0
        subtitleLabel.adjustsFontForContentSizeCategory = true

        let textStack = UIStackView(arrangedSubviews: [titleLabel, subtitleLabel])
        textStack.axis = .vertical
        textStack.spacing = 2

        var trailingViews: [UIView] = []
        if let accessoryText = accessoryText {
            trailingViews.append(makeAccessoryBadge(accessoryText))
        }
        let chevron = UIImageView(image: UIImage(systemName: "chevron.right"))
        chevron.tintColor = .tertiaryLabel
        chevron.setContentHuggingPriority(.required, for: .horizontal)
        trailingViews.append(chevron)

        let row = UIStackView(arrangedSubviews: [icon, textStack] + trailingViews)
        row.axis = .horizontal
        row.spacing = Spacing.m
        row.alignment = .center
        row.isUserInteractionEnabled = false
        row.translatesAutoresizingMaskIntoConstraints = false
        addSubview(row)
        NSLayoutConstraint.activate([
            row.topAnchor.constraint(equalTo: topAnchor, constant: 14),
            row.bottomAnchor.constraint(equalTo: bottomAnchor, constant: -14),
            row.leadingAnchor.constraint(equalTo: leadingAnchor, constant: 16),
            row.trailingAnchor.constraint(equalTo: trailingAnchor, constant: -16),

            icon.widthAnchor.constraint(equalToConstant: 28),
            icon.heightAnchor.constraint(equalToConstant: 28)
        ])

        let tap = UIButton(type: .system)
        tap.translatesAutoresizingMaskIntoConstraints = false
        tap.addTarget(self, action: #selector(handleTap), for: .touchUpInside)
        addSubview(tap)
        NSLayoutConstraint.activate([
            tap.topAnchor.constraint(equalTo: topAnchor),
            tap.bottomAnchor.constraint(equalTo: bottomAnchor),
            tap.leadingAnchor.constraint(equalTo: leadingAnchor),
            tap.trailingAnchor.constraint(equalTo: trailingAnchor)
        ])
    }

    @available(*, unavailable)
    required init?(coder: NSCoder) { fatalError() }

    @objc private func handleTap() { action() }

    private func makeAccessoryBadge(_ text: String) -> UIView {
        let label = UILabel()
        label.text = text
        label.font = UIFont.preferredFont(forTextStyle: .caption2).withWeight(.semibold)
        label.textColor = .secondaryLabel
        label.adjustsFontForContentSizeCategory = true

        let pill = UIView()
        pill.backgroundColor = UIColor.secondaryLabel.withAlphaComponent(0.12)
        pill.layer.cornerRadius = 8
        pill.layer.cornerCurve = .continuous
        pill.translatesAutoresizingMaskIntoConstraints = false

        label.translatesAutoresizingMaskIntoConstraints = false
        pill.addSubview(label)
        NSLayoutConstraint.activate([
            label.topAnchor.constraint(equalTo: pill.topAnchor, constant: 4),
            label.bottomAnchor.constraint(equalTo: pill.bottomAnchor, constant: -4),
            label.leadingAnchor.constraint(equalTo: pill.leadingAnchor, constant: 8),
            label.trailingAnchor.constraint(equalTo: pill.trailingAnchor, constant: -8)
        ])
        return pill
    }
}

/// Monospaced append-only log used by example screens to surface SDK
/// lifecycle callbacks. Always scrolls to the latest line.
final class LogView: UIView {

    private let textView = UITextView()

    override init(frame: CGRect) {
        super.init(frame: frame)
        backgroundColor = .systemGroupedBackground
        layer.cornerRadius = 8
        layer.cornerCurve = .continuous
        clipsToBounds = true

        textView.font = .monospacedSystemFont(ofSize: 11, weight: .regular)
        textView.textColor = .label
        textView.backgroundColor = .clear
        textView.isEditable = false
        textView.alwaysBounceVertical = true
        textView.textContainerInset = UIEdgeInsets(top: 8, left: 8, bottom: 8, right: 8)
        textView.translatesAutoresizingMaskIntoConstraints = false
        addSubview(textView)
        NSLayoutConstraint.activate([
            textView.topAnchor.constraint(equalTo: topAnchor),
            textView.bottomAnchor.constraint(equalTo: bottomAnchor),
            textView.leadingAnchor.constraint(equalTo: leadingAnchor),
            textView.trailingAnchor.constraint(equalTo: trailingAnchor),
            textView.heightAnchor.constraint(greaterThanOrEqualToConstant: 160)
        ])
    }

    @available(*, unavailable)
    required init?(coder: NSCoder) { fatalError() }

    func append(_ line: String) {
        let stamp = DateFormatter.localizedString(from: Date(), dateStyle: .none, timeStyle: .medium)
        textView.text.append(contentsOf: "[\(stamp)] \(line)\n")
        let end = NSRange(location: (textView.text as NSString).length, length: 0)
        textView.scrollRangeToVisible(end)
    }

    func clear() { textView.text = "" }
}
